File: src/secondary/help.lua
--- Provide help on commands and general usage.
-- @module help

local version = "0.1.14"
local progname = "tman"

local function show_usage()
    io.stdout:write(([[
Usage: %s COMMAND [OPTION] [ID]

List of available commands:

System:
  archive - backup and restore metadata
  config  - config your workflow
  get     - get parameters like curr, prev tasks, size, etc
  init    - init util structure

Basic:
  cat     - cat task info
  env     - define or display environments
  list    - list tasks
  prev    - switch to previous task
  use     - mark a task as current

Change:
  add     - add new task
  del     - delete task
  pack    - pack commits in repos for review.
  set     - set task units
  sync    - synchronize task struct, git branch, repos, etc

Info:
  help    - display help information about %s
  ver     - show version

'%s help COMMAND' to get detailed info about command.
'%s help %s' to get info about util itself.
]]):format(progname, progname, progname, progname, progname))
end

local cmds = {
    {
        name = progname,
        desc = [[
tman - terminal task manager.

Util to switch between tasks, git branches and make the workflow easy and
less confusing. You no longer need to memerize task IDs, branche names,
cd between directories, lose task notes, but instead look
for 'em in nice structured way.
]],
    },
    {
        name = "init",
        desc = [[
Usage: tman init
Initialize util.
]],
    },
    {
        name = "archive",
        desc = [[
Usage: tman archive OPTION
Make backup and restore util data.

Options:
    -R          include repos into archive (default: false)
    -b FILE     create backup (default extension: .tar)
    -h          Show this help message.
    -r FILE     restore from archive
]],
    },
    {
        name = "config",
        desc = [[
Usage: tman config OPTION
Work with config files.

Options:
    -b      - set base path. Prefix is user home.
    -i      - set install path
              questionable, cuz util gotta change path in .shellrc
    -s      - show config: system & user
]],
    },
    {
        name = "add",
        desc = [[
Usage: tman add [OPTION] ID
Add new task.

Options:
    -d      set task description (under development)
    -h      show this help message.
    -p      set task priority. Values: [highest|high|mid|low|lowest]
    -t      set task type. Values: [bugfix|hotfix|feature]
]],
    },
    {
        name = "del",
        desc = [[
Usage: tman del [ID]
Delete task.
ID - default is current task.
]],
    },
    {
        name = "set",
        desc = [[
Usage: tman set OPTION [ID]
Amend task items.
ID - default is current task.

Options:
    -i      set task ID
    -d      set task description
    -l      set task link
    -p      set task priority. Values: [highest|high|mid|low|lowest]
    -t      set task type. Values: [bugfix|hotfix|feature]
]],
    },
    {
        name = "sync",
        desc = [[
Usage: tman sync OPTION [ID]
Update task repos, structure, etc. Default: current task.
With no option applied jump to task directory.

Notes: Add support for other task cuz another task might be broken so
      repair's need (if task branch's deleted)

Options:
    -s      - synchronize task structure
    -r      - git pull from remote repo
    -t      - synchronize task status (active, completed)
]],
    },

    {
        name = "use",
        desc = [[
Usage: tman use ID
Switch to specified task. Use `tman list` to see existing tasks.

Notes:
    ID - taks ID. Default is current task.
]],
    },
    {
        name = "get",
        desc = [[
Usage: tman get PARAM
Get parameters like curr, prev tasks, current env, etc

Notes:
PARAM can have one of the next values
    env     - current env
    curr    - current task ID
    prev    - previous task ID
]],
    },
    {
        name = "env",
        desc = [[
Usage: tman env COMMAND [ENVNAME]
Define or display environments.

ENVNAME - environment name

COMMANDs:
    add     - add new environment
    curr    - show current environment
    del     - delete an environment
    list    - list environments
    prev    - switch to previous environment
    use     - use environment
]],
    },
    {
        name = "prev",
        desc = [[
Usage: tman prev [ENVNAME]
Switch to previous task if set.

ENVNAME - environment name. Under development (pro'ly).

Options:
    -h   Show this help message.
]],
    },
    {
        name = "pack",
        desc = [[
Usage: tman pack OPTION [ID]
Pack commits in repos for review.

Options:
    -c      - create commit (default)
    -m      - run commands from the Makefile
    -p      - push branch to remote repo
]],
    },
    {
        name = "list",
        desc = [[
Usage: tman list [OPTION]
List task IDs with description.

Options:
    -c   List only complete tasks.
    -a   List only active tasks (default).
    -A   List all tasks: active and complete.
    -h   Show this help message.

Notes:
    *   Marks current task.
    -   Makrs previous task.
]],
    },
    {
        name = "cat",
        desc = [[
Usage: tman cat [OPTION] [ID]
Show task units.
ID - task ID (default is current task).

Options:
    -k   cat specific task unit
    -h   Show this help message.
]],
    },
    {
        name = "help",
        desc = [[
Usage: tman help [COMMAND]
Show help message. If `COMMAND' applied then show info about command.
]],
    },
    {
        name = "ver",
        desc = [[
Usage: tman ver
Show tman version.
]],
    },
}

--- Get general and detailed info about command.
-- @param cmdname command name to get info about
-- @return status code: 0 - ok, otherwise error code
local function help_usage(cmdname)
    local errmsg = "%s: no such command '%s'. Use '%s help' for more info.\n"

    if not cmdname then
        show_usage()
        return 0
    end

    for _, cmd in ipairs(cmds) do
        if cmd.name == cmdname then
            io.stdout:write(cmd.desc)
            return 0
        end
    end
    io.stderr:write(errmsg:format(progname, cmdname, progname))
    return 1
end

return {
    usage = help_usage,
    version = version,
    progname = progname,
}

File: src/help.c
#include <stdio.h>
#include <string.h>
#include "common.h"
#include "help.h"

#define TAGSYSTEM   "bin-system"
#define TAGBASIC    "bin-basic"
#define TAGMISC     "bin-misc"
#define TAGINFO     "bin-info"

struct help help[] = {
    {
        .tag = TAGSYSTEM,
        .name  = "cfg",
        .sdesc = "manipulate system configuration",
        .desc  = "cfg description"
    },
    {
        .tag = TAGSYSTEM,
        .name  = "chk",
        .sdesc = "check system config and task units",
        .desc  = "cfg description"
    },
    {
        .tag = TAGSYSTEM,
        .name  = "init",
        .sdesc = "init directory structure",
        .desc  = "init description"
    },
    {
        .tag = TAGSYSTEM,
        .name  = "pgm",
        .sdesc = "plugin manager",
        .desc  = "pgm description"
    },

    {
        .tag = TAGBASIC,
        .name  = "add",
        .synop = "Usage: " PROGRAM " add [OPTION]... ID...",
        .sdesc = "Add a new task into environment",
        .desc  = "With no option `-e' add a new task to current environment.\n"
                 "If current environment is not set throw then an error.\n\n"
                 "  -e      specify an environment to add a task to\n"
                 "  -f      do not output error message if task aleady exists\n"
                 "  -h      show this help and exis"
    },
    {
        .tag = TAGBASIC,
        .name  = "col",
        .synop = "Usage: " PROGRAM " col [OPTION]... ID..\n"
                 "Try '" PROGRAM " help col' for more info.",
        .sdesc = "move task to specific column",
        .desc  = "col description",
    },
    {
        .tag = TAGBASIC,
        .name  = "del",
        .synop = "Usage: " PROGRAM " del [OPTION]... ID..\n"
                 "Try '" PROGRAM " help del' for more info.",
        .sdesc = "Delete the task from the environment",
        .desc  = "del description",
    },
    {
        .tag = TAGBASIC,
        .name  = "prev",
        .synop = PROGRAM " prev",
        .sdesc = "switch to previous task",
        .desc  = "prev description",
    },
    {
        .tag = TAGBASIC,
        .name  = "set",
        .synop = PROGRAM " set OPTION... [ID]... [ENV]",
        .sdesc = "set task unit values",
        .desc  = "set description",
    },
    {
        .tag = TAGBASIC,
        .name  = "sync",
        .synop = PROGRAM " sync",
        .sdesc = "synchronize task",
        .desc  = "sync description",
    },
    {
        .tag = TAGBASIC,
        .name  = "use",
        .synop = PROGRAM " use ID [ENV]",
        .sdesc = "switch to specific task",
        .desc  = "use description",
    },

    {
        .tag = TAGMISC,
        .name  = "env",
        .synop = PROGRAM " env SUBCMD [OPTION] NAME",
        .sdesc = "manipulate environments",
        .desc  = "env description",
    },
    {
        .tag = TAGMISC,
        .name  = "move",
        .synop = PROGRAM " move [OPTION] ID DST SRC",
        .sdesc = "move task to another environment",
        .desc  = "move description",
    },

    {
        .tag = TAGINFO,
        .name  = "cat",
        .synop = PROGRAM " cat ID... [ENV]",
        .sdesc = "cat task info",
        .desc  = "cat description",
    },
    {
        .tag = TAGINFO,
        .name  = "help",
        .synop = PROGRAM " help [CMD]",
        .sdesc = "show this help message and exit",
        .desc  = "help description",
    },
    {
        .tag = TAGINFO,
        .name  = "list",
        .synop = PROGRAM " list [ENV]",
        .sdesc = "list environment task",
        .desc  = "help description",
    },
    {
        .tag = TAGINFO,
        .name  = "ver",
        .synop = PROGRAM " ver",
        .sdesc = "show version and exit",
        .desc  = "ver description",
    },
};

static size_t helpsize = sizeof(help) / sizeof(help[0]);

int help_list_commands(void)
{
    printf("System:\n");
    for (int i = 0; i < helpsize; ++i)
        if (strcmp(help[i].tag, TAGSYSTEM) == 0)
            printf("  %-6s - %s\n", help[i].name, help[i].sdesc);

    printf("\nBasic:\n");
    for (int i = 0; i < helpsize; ++i)
        if (strcmp(help[i].tag, TAGBASIC) == 0)
            printf("  %-6s - %s\n", help[i].name, help[i].sdesc);

    printf("\nMisc:\n");
    for (int i = 0; i < helpsize; ++i)
        if (strcmp(help[i].tag, TAGMISC) == 0)
            printf("  %-6s - %s\n", help[i].name, help[i].sdesc);

    printf("\nInfo:\n");
    for (int i = 0; i < helpsize; ++i)
        if (strcmp(help[i].tag, TAGINFO) == 0)
            printf("  %-6s - %s\n", help[i].name, help[i].sdesc);
    return 1;
}

int help_usage(const char *cmd)
{
    for (int i = 0; i < helpsize; ++i) {
        if (strcmp(help[i].name, cmd) == 0) {
            printf("%s\n", help[i].synop);
            printf("%s\n", help[i].sdesc);
            printf("Try '" PROGRAM " help %s' for more info.\n", cmd);
            return 1;
        }
    }
    return elog(1, "cannot access '%s': command not found", cmd);
}

int help_lookup(const char *cmd)
{
    if (cmd == NULL)
        return help_list_commands();
    for (int i = 0; i < helpsize; ++i) {
        if (strcmp(help[i].name, cmd) == 0) {
            printf("%s\n", help[i].synop);
            printf("%s\n\n", help[i].sdesc);
            printf("%s\n", help[i].desc);
            return 1;
        }
    }
    return elog(1, "cannot access '%s': command not found", cmd);
}
